contrib/packs/actions/pack_mgmt/download.py

```diff
@@ -51,46 +52,63 @@ def run(self, packs, abs_repo_base, verifyssl=True):
-                pack_name = self._get_pack_name(abs_local_path)
                 try:
+                    user_home = os.path.expanduser('~')
+                    abs_local_path = os.path.join(user_home, temp_dir)
+                    self._clone_repo(temp_dir=abs_local_path, repo_url=pack_url,
+                                     verifyssl=verifyssl, ref=pack_version)
+                    pack_name = self._get_pack_name(abs_local_path)
                     result[pack_name] = self._move_pack(abs_repo_base, pack_name, abs_local_path)
                 finally:
                     self._cleanup_repo(abs_local_path)
 
         return self._validate_result(result=result, repo_url=pack_url)
 
     @staticmethod
-    def _clone_repo(temp_dir, repo_url, verifyssl=True, ref=None):
-        user_home = os.path.expanduser('~')
-        abs_local_path = os.path.join(user_home, temp_dir)
-
+    def _clone_repo(temp_dir, repo_url, verifyssl=True, ref='master'):
         # Switch to non-interactive mode
         os.environ['GIT_TERMINAL_PROMPT'] = '0'
 
         # Disable SSL cert checking if explictly asked
         if not verifyssl:
             os.environ['GIT_SSL_NO_VERIFY'] = 'true'
 
-        if not ref:
-            ref = 'master'
-
         # Clone the repo from git; we don't use shallow copying
         # because we want the user to work with the repo in the
         # future.
-        repo = Repo.clone_from(repo_url, abs_local_path, branch='master')
-
-        if not DownloadGitRepoAction._ref_exists(repo, ref):
-            if re.match(SEMVER_REGEX, ref) and DownloadGitRepoAction._ref_exists(repo, "v%s" % ref):
-                ref = "v%s" % ref
-            else:
-                raise ValueError("\"%s\" is not a valid ref in %s." % (ref, repo_url))
+        repo = Repo.clone_from(repo_url, temp_dir)
+
+        # Try to match the reference to a commit hash, a tag, or "master"
+        gitref = DownloadGitRepoAction._get_gitref(repo, ref)
+
+        # Try to match the reference to a "vX.Y.Z" tag
+        if not gitref and re.match(SEMVER_REGEX, ref):
+            gitref = DownloadGitRepoAction._get_gitref(repo, "v%s" % ref)
+
+        # Try to match the reference to a branch name
+        if not gitref:
+            gitref = DownloadGitRepoAction._get_gitref(repo, "origin/%s" % ref)
+
+        # Giving up ¯\_(ツ)_/¯
+        if not gitref:
+            raise ValueError(
+                "\"%s\" is not a valid version, hash, tag, or branch in %s." % (ref, repo_url)
+            )
+
+        # We're trying to figure out which branch the ref is actually on,
+        # since there's no direct way to check for this in git-python.
+        branches = repo.git.branch('--color=never', '--all', '--contains', gitref.hexsha)
+        branches = branches.replace('*', '').split()
+        if 'master' not in branches:
+            branch = branches[0]
+            repo.git.checkout('--track', branches[0])
```

Comment:
`gitpython` is quite horrible (also I don't like the absence of a plumbing command for `git branch --contains`, but yeah, this works. :)